File: benchmarks/benchmarks.py
import sys
import os
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))

import numpy as np
import optim_engine as opt
import scipy.optimize as sciop
from scipy.optimize import dual_annealing as da
import time
from typing import Callable, Any

def benchmark(name: str, func: Callable, runs: int = 10) -> tuple[float, Any]:

    runtimes = []
    res = None

    for _ in range(runs):

        start = time.perf_counter()
        res = func()
        end = time.perf_counter()

        runtimes.append((end - start) * 1000)

    av_time = sum(runtimes) / len(runtimes)

    return av_time, res

# Gradient Descent

x_0 = np.array([5.0, 5.0])
sphere = opt.Sphere(2)
gd_config = opt.GDConfig()
gd = opt.GradientDescent(gd_config)
fn_gd = lambda: gd.optimize(sphere, x_0)
av_time_gd, res_gd_ = benchmark("Gradient Descent", fn_gd)

fn_gd_sci = lambda: sciop.minimize(
    fun = lambda x: float(np.dot(x, x)),
    x0 = np.array([5.0, 5.0]),
    jac = lambda x: 2.0 * x,
    method = 'CG'
)

av_time_scigd, res_scigd = benchmark("SciPy GD", fn_gd_sci)

# Newton

newton_config = opt.NewtonConfig()
newton = opt.Newton(newton_config)
fn_newton = lambda: newton.optimize(sphere, x_0)
av_time_newt, res_newt_ = benchmark("Newton Descent", fn_newton)

fn_newt_sci = lambda: sciop.minimize(
    fun = lambda x: float(np.dot(x, x)),
    x0 = x_0,
    jac = lambda x: 2.0 * x,
    hess = lambda x: 2.0 * np.eye(len(x)),
    method = "Newton-CG"
)

av_time_scinew, res_scinew = benchmark("SciPy Newton", fn_newt_sci)

# Hill Climbing
'''
Since SciPy doesn't have a direct Hill Climbing method, 
the implemented C++ method is compared against SciPy's 
Nelder-Mead Method.
'''
hc_config = opt.HCConfig()
hc_config.max_epochs = 100000
hc_config.step_size = 0.1
hc = opt.HC(hc_config)
fn_hc = lambda: hc.optimize(sphere, x_0)
av_time_hc, res_hc_ = benchmark("Hill Climbing", fn_hc)

fn_nm_sci = lambda: sciop.minimize(
    fun = lambda x: float(np.dot(x, x)),
    x0 = x_0,
    method = "Nelder-Mead"
)

av_time_scinm, res_scinm = benchmark("SciPy Nelder-Mead", fn_nm_sci)

# Simulated Annealing
sa_config = opt.SAConfig()
sa = opt.SA(sa_config)
fn_sa = lambda: sa.optimize(sphere, x_0)
av_time_sa, res_sa_ = benchmark("Simulated Annealing", fn_sa)

fn_sa_sci = lambda: da(
    func = lambda x: float(np.dot(x, x)),
    bounds = [(-10, 10), (-10, 10)],
    x0 = x_0
)

av_time_scisa, res_scisa = benchmark("SciPy Dual Annealing", fn_sa_sci)

print("Results:")
print(f"{'Optimizer':<20} {'Time(ms)':<12} {'Iters':<10} {'f(x*)'}\n")
print(f"{'Gradient Descent':<20} {av_time_gd:<12.3f} {res_gd_.iterations:<10} {res_gd_.optimal_f:.2e}\n")
print(f"{"SciPy GD":<20} {av_time_scigd:<12.3f} {res_scigd.nit:<10} {res_scigd.fun:.2e}\n")
print(f"{'Newton Descent':<20} {av_time_newt:<12.3f} {res_newt_.iterations:<10} {res_newt_.optimal_f:2e}\n")
print(f"{"SciPy Newton":<20} {av_time_scinew:<12.3f} {res_scinew.nit:<10} {res_scinew.fun:.2e}\n")
print(f"{'Hill Climbing':<20} {av_time_hc:<12.3f} {res_hc_.iterations:<10} {res_hc_.optimal_f:.2e}\n")
print(f"{"SciPy Nelder-Mead":<20} {av_time_scinm:<12.3f} {res_scinm.nit:<10} {res_scinm.fun:.2e}\n")
print(f"{'Simulated Annealing':<20} {av_time_sa:<12.3f} {res_sa_.iterations:<10} {res_sa_.optimal_f:.2e}\n")
print(f"{"SciPy Dual Annealing":<20} {av_time_scisa:<12.3f} {res_scisa.nit:<10} {res_scisa.fun:.2e}\n")

File: benchmarks/benchmarks.cpp
#include <iostream>
#include <chrono>
#include <iomanip>
#include <string>
#include <Eigen/Dense>
#include "umbrella_header.hpp"

template<typename OptimizerType, typename FunctionType>
void run_benchmark(const std::string &name,
                   const std::string &function_name,
                   OptimizerType &optimizer,
                   FunctionType &function,
                   const Eigen::VectorXd &x0,
                   int runs = 10)
{
    double total_ms{};
    OptimResult result;

    for(int i{}; i < runs; ++i)
    {
        auto start = std::chrono::high_resolution_clock::now();
        result = optimizer.optimize(function, x0);
        auto end = std::chrono::high_resolution_clock::now();
        total_ms += std::chrono::duration<double, std::milli>(end - start).count();
    }

    double avg_ms{total_ms / runs};

    std::cout << std::left
              << std::setw(22) << name
              << std::setw(15) << function_name
              << std::setw(12) << std::fixed << std::setprecision(3) << avg_ms
              << std::setw(10) << result.iterations
              << std::setw(12) << std::scientific << std::setprecision(2) << result.optimal_f
              << std::setw(12) << (result.status == Status::CONVERGED ? "CONVERGED" :
                                   result.status == Status::STAGNATED ? "STAGNATED" :
                                   result.status == Status::FAILED ? "FAILED" : "MAX_EPOCHS")
              << '\n';
}

int main()
{
    Eigen::VectorXd xs_0(2);
    xs_0 << 5.0, 5.0;
    Eigen::VectorXd xr_0(2);
    xr_0 << 1.0, -1.0;
    Eigen::VectorXd x_rast(4);
    x_rast << 2.0, 3.0, 4.0, 5.0;

    std::cout << "\n";
    std::cout << "Benchmark:\n";
    std::cout << std::left
              << std::setw(22) << "Optimizer"
              << std::setw(15) << "Function"
              << std::setw(12) << "Time(ms)"
              << std::setw(10) << "Epochs"
              << std::setw(12) << "f(x_opt)"
              << std::setw(12) << "Status" << '\n';

    Sphere sphere(2);
    Rosenbrock rosenbrock(2);
    Rastrigin rastrigin(4);

    // Benchmark for GD
    GDConfig gd_config;
    gd_config.learning_rate = 0.01;
    gd_config.max_epochs = 10000;
    GradientDescent gd(gd_config);
    run_benchmark("Gradient Descent", "Sphere", gd, sphere, xs_0);
    run_benchmark("Gradient Descent", "Rosenbrock", gd, rosenbrock, xr_0);
    run_benchmark("Gradient Descent", "Rastrigin", gd, rastrigin, x_rast);

    //Benchmark for Newton
    NewtonConfig newton_config;
    Newton newton(newton_config);
    run_benchmark("Newton Descent", "Sphere", newton, sphere, xs_0);
    run_benchmark("Newton Descent", "Rosenbrock", newton, rosenbrock, xr_0);
    run_benchmark("Newton Descent", "Rastrigin", newton, rastrigin, x_rast);

    //Benchmark for Hill Climbing
    HCConfig hc_config;
    hc_config.max_epochs = 100000;
    hc_config.step_size = 0.1;
    HC hc(hc_config);
    run_benchmark("Hill Climbing", "Sphere", hc, sphere, xs_0);
    run_benchmark("Hill Climbing", "Rosenbrock", hc, rosenbrock, xr_0);
    run_benchmark("Hill Climbing", "Rastrigin", hc, rastrigin, x_rast);

    //Benchmark for Simulated Annealing
    SAConfig sa_config;
    SA sa(sa_config);
    run_benchmark("Simulated Annealing", "Sphere", sa, sphere, xs_0);
    run_benchmark("Simulated Annealing", "Rosenbrock", sa, rosenbrock, xr_0);
    run_benchmark("Simulated Annealing", "Rastrigin", sa, rastrigin, x_rast);
}